### PR TITLE
Control operator memory by absolute values in resource group mode

### PR DESCRIPTION
**Pull request by zhangh43**
Old memory_spill_ratio only supports percentage format,
which is hard for user to config the operator(join, sort etc.)
memory. We add absolute value format to make memory_spill_ratio
have the similar behavior like GUC statement_mem in resource
queue mode.

Also change the default values of resource group, e.g. use 128MB
as memory_spill_ratio to make the migration from resource queue
to resource group more smoothly.

Example to set absolute value for memory_spill_ratio:
```
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=0, memory_spill_ratio='100MB');
```

Co-authored-by: Ning Yu <nyu@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
Added a `WIP` tag, we still have something to do:
- retire the `proposed` attributed;
- bump catalog version;

This PR will also be back ported to 6X and 5X, for 6X it will introduce a catalog change, for 5X we will only back port the non-catalog-change parts.  Binary swap tests will be added for both branches.

File: src/backend/cdb/cdbvars.c [hunk 1]
```diff
@@ -699,6 +700,33 @@ gpvars_show_gp_resource_manager_policy(void)
 	}
 }
 
+bool
+gpvars_check_memory_spill_ratio(const char **newval, void **extra, GucSource source)
```
Comment (pivotal-ning-yu) [inline src/backend/cdb/cdbvars.c]: let's move these hooks to guc_gp.c, so we could make them `static` and remove the `gpvars_` prefix.  Maybe we could put them after `assign_gp_resource_group_bypass()`.

By doing so we could prevent touching cdbvars.c and guc.h .

File: src/include/utils/resgroup.h [hunk 2]
```diff
@@ -201,7 +201,8 @@ extern void CpusetDifference(char *cpuset1, const char *cpuset2, int len);
 extern bool CpusetIsEmpty(const char *cpuset);
 extern void SetCpusetEmpty(char *cpuset, int cpusetSize);
 extern bool EnsureCpusetIsAvailable(int elevel);
-
+extern int32 ResGroupMemorySpillFromStr(const char *str, const char *prop);
+extern void ResGroupMemorySpillToStr(int32 value, char *buf, int bufsize);
```
Comment (pivotal-ning-yu) [inline src/include/utils/resgroup.h]: Let's move these definitions between `ResGroupGetSegmentNum()` and `CpusetToBitset()`.

File: src/backend/commands/resgroupcmds.c [hunk 3]
```diff
@@ -1588,3 +1603,97 @@ checkCpusetSyntax(const char *cpuset)
...
+	int32 value;
+
+	/* if memory spill ratio is percentile value */
+	if (parse_int(str, &value, 0, NULL))
+	{
+		if (value < RESGROUP_MIN_MEMORY_SPILL_RATIO ||
+			value > RESGROUP_MAX_MEMORY_SPILL_RATIO)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("memory_spill_ratio range is [%d, %d]",
```
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: Let's reword the error message like this:

```
memory_spill_ratio in percentage format must be in the range [0, 100]
```

File: src/backend/commands/resgroupcmds.c [hunk 4]
```diff
@@ -1588,3 +1603,97 @@ checkCpusetSyntax(const char *cpuset)
...
+ *   percentage;
+ * - the absolute value format is in "number unit" format, where "unit" can be
+ *   "kB", "MB", "GB" or "TB", the "number" must be > 0, it is converted to
+ *   chunks, return (-1 * chunks);
+ */
+int32
+ResGroupMemorySpillFromStr(const char *str, const char *prop)
+{
+
+	int32 value;
```
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: Let's remove the empty line, correct the space between `int32` and `value`, we could also convert the arg `prop` to a local const var:

```C
int32
ResGroupMemorySpillFromStr(const char *str)
{
    const char *prop = "memory_spill_ratio";
    int32       value;
```

File: src/backend/commands/resgroupcmds.c [hunk 5]
```diff
@@ -1588,3 +1603,97 @@ checkCpusetSyntax(const char *cpuset)
...
+ * Convert memory_spill_ratio to a string.
+ *
+ * Refer to ResGroupMemorySpillFromStr() for details of the string format.
+ */
+void
+ResGroupMemorySpillToStr(int32 value, char *buf, int bufsize)
+{
+#define KB_PER_MB (1024)
+#define KB_PER_GB (1024*1024)
+#define KB_PER_TB (1024*1024*1024)
```
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: Let's move these macros after the comment `Below logic is derived from _ShowOption()` as these macros are also derived from that function.

File: src/backend/commands/resgroupcmds.c [hunk 6]
```diff
@@ -1588,3 +1603,97 @@ checkCpusetSyntax(const char *cpuset)
...
+	{
+		/* The value is in the percentage format */
+		snprintf(buf, bufsize, "%d", value);
+	}
+	else
+	{
+		/* The value is in the absolute value format */
+
+		/* Below logic is derived from _ShowOption() */
+		int64		result = -(VmemTracker_ConvertVmemChunksToMB(value) << 10);
```
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: This line is too line, let's separate the defination and the assignment.

```C
        int64       result;
        const char *unit;

        result = -(VmemTracker_ConvertVmemChunksToMB(value) << 10);
```

File: src/backend/commands/resgroupcmds.c (hunk 4 above)
Comment (zhangh43) [reply src/backend/commands/resgroupcmds.c]: fixed

File: src/test/isolation2/input/resgroup/enable_resgroup.source [hunk 7]
```diff
@@ -63,4 +63,4 @@ DROP FUNCTION adjust_memory_limit(bigint);
 0: ALTER RESOURCE GROUP admin_group SET concurrency 2;
 -- in later cases we will SHOW memory_spill_ratio as first command
 -- to verify that it can be correctly loaded even for bypassed commands
-0: ALTER RESOURCE GROUP admin_group SET memory_spill_ratio 10;
+0: ALTER RESOURCE GROUP admin_group SET memory_spill_ratio '128 MB';
```
[the remark this answers is not in the document]
Comment (zhangh43) [reply src/test/isolation2/input/resgroup/enable_resgroup.source]: make sense

**Comment by zhangh43** [comment]
The following tasks have finished.
- retire the `proposed` attributed;
- bump catalog version;

**Pull request merged**